### DISCUSSION TITLE
Custom Target that compiles into a subdirectory

### DISCUSSION THREAD

**Discussion by tmayoff** [mesonbuild/meson#10488] [Q&A]
I'm trying to compile some custom sources from .idl that compiles some files in the current build dir, as well as a subdirectory. This isn't possible with the `custom_target` `custom_target keyword argument "output" Output 'AnyTypeCode/@BASENAME@A.h' must not contain a path segment.` Is there a way to accomplish this properly?

Example of what I'm talking about
```
/
├── build <-- Generated
│   ├── AnyTypeCode
│   │   ├── GIOPA.h
│   │   └── GIOPA.cpp
│   ├── GIOPC.h
│   └── GIOPC.cpp
└── src <-- Original
    └── GIOP.pidl 
```

**Comment by eli-schwartz** [comment]
This is the stated purpose of `generator(...).process(...)`: https://mesonbuild.com/Reference-manual_returned_generator.html#generatorprocess

Using the `preserve_path_from: meson.current_source_dir()` kwarg, all the generated headers (generated into the executable/library's private dir) will preserve the path structure.

**Reply by tmayoff** [reply]
I don't think this quite solves my issue unless I'm misunderstanding. This is the structure that needs to be generated

```
/
├── build <-- Generated
│   ├── AnyTypeCode
│   │   ├── GIOPA.h
│   │   └── GIOPA.cpp
│   ├── GIOPC.h
│   └── GIOPC.cpp
└── src <-- Original
    └── GIOP.pidl 
```

The files do indeed get generated in the correct location but adding them to the library sources to be compiled is the issue. The files specified in the `output` argument of either generator or custom_target can't contain a path segment to specify the AnyTypeCode subdir

**Reply by eli-schwartz** [reply]
Oh hmm, my understanding was that there's a common case of generated headers, in which you have `'foo/bar.gen', 'foo/bar/baz.gen'` and you want to generate `'foo/bar.h', 'foo/bar.cpp', 'foo/bar/baz.h', 'foo/bar/baz.cpp'`. This much plays nicely with preserve_path_from.

**Reply by tmayoff** [reply]
Oh yeah sorry my original question was poorly explained, do you know if this is possible to do simply?

**Reply by mitchgrout** [reply]
Bumping this; I have a similar need to generate source/headers from .proto files while maintaining a preserved folder structure, but the key problem is that `generator.process()` produces a target that acts differently from a standard `custom_target`, so if the generated headers get used in multiple build targets, the headers will be generated multiple times.

Are there any alternatives to this?

**Reply by mitchgrout** [reply]
What does the "no path segments allowed" check aim to mitigate? I'm wanting to create a custom packaging around something like https://protobuf.dev/reference/protobuf/google.protobuf/ which will internally contain directives like `import "google.protobuf.[module]"`, so it is convenient for me to declare something like...

```meson
custom_target(
  input: fs.copyfile('google/protobuf/api.proto'),
  output: ['google/protobuf/api.pb.h', 'google/protobuf/api.pb.cc'],
  command: ['protoc', '--cpp_out=.', '@INPUT@'],
)
```

AFAIK `custom_target` should not assume it is in a particular directory, but I don't think there's many alternatives to this?

**Reply by eli-schwartz** [reply]
See https://github.com/mesonbuild/meson/issues/2320
